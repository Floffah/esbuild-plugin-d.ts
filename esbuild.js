const {build} = require("esbuild");

const dev = process.argv.includes("--dev") || process.argv.includes("-d");

build({
    bundle: true,
    target: "node16.0",
    platform: "node",
    watch: dev,
    color: true,
    logLevel: "info",
    minify: !dev,
    minifyIdentifiers: !dev,
    minifySyntax: !dev,
    minifyWhitespace: !dev,
    format: "cjs",
    entryPoints: ["./src/index.ts"],
    outfile: "./dist/index.cjs",
    external: ["esbuild", "typescript", "chalk"]
})
