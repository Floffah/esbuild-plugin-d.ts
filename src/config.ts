import ts from "typescript";
import { readFileSync } from "fs";

export function getTSConfig(conf?: string, wd = process.cwd()): {loc: string, conf: any} {
    const f = ts.findConfigFile(wd, ts.sys.fileExists, conf);
    if (!f) throw "No config file found";
    const c = ts.readConfigFile(f, (path) => readFileSync(path, "utf-8"));
    if (c.error) throw c.error;
    else return {loc: f, conf: c.config};
}

export interface DTSPluginOpts {
    outDir?: string;
}
