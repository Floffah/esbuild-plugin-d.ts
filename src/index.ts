export * from "./lib";
export * from "./types";
export * from "./plugin";
export { dtsPlugin as default } from "./plugin";
