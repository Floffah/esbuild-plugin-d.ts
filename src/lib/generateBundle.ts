import { generateDtsBundle } from "dts-bundle-generator";
import { randomBytes } from "node:crypto";
import { rmSync, writeFileSync } from "node:fs";
import { dirname, sep as pathSeparator, resolve } from "node:path";
import ts from "typescript";

function getHighestCommonDirectory(paths: string[]): string {
    if (paths.length === 0) {
        return "";
    } else if (paths.length === 1) {
        return dirname(paths[0]);
    }

    const [firstPath, ...otherPaths] = paths.map((p) => p.split(pathSeparator));
    let commonParts = firstPath;

    for (const currentPath of otherPaths) {
        commonParts = commonParts.slice(
            0,
            commonParts.findIndex((part, i) => part !== currentPath[i]) ||
                commonParts.length,
        );
    }

    return commonParts.join(pathSeparator);
}

export function generateBundle(
    entryPoints: string[],
    compilerOptions: ts.CompilerOptions,
    tsconfigPath?: string,
    originalConfig: any = {},
) {
    const commonOutDir = getHighestCommonDirectory(entryPoints);

    const relativeDeclarationPaths = entryPoints.map((entry) =>
        entry.replace(commonOutDir + "/", "").replace(/\.tsx?$/, ".d.ts"),
    );
    const postbundleOutDir = resolve(compilerOptions.declarationDir!, "..");

    let shouldDeleteTsConfig = false;
    if (!tsconfigPath) {
        const tempid = randomBytes(20).toString("hex");

        tsconfigPath = resolve(process.cwd(), `tsconfig.${tempid}.json`);

        writeFileSync(
            tsconfigPath,
            JSON.stringify({
                ...originalConfig,
                compilerOptions,
                include: entryPoints,
            }),
        );

        shouldDeleteTsConfig = true;
    }

    const bundles = generateDtsBundle(
        relativeDeclarationPaths.map((path) => ({
            filePath: resolve(compilerOptions.declarationDir!, path),
        })),
        {
            preferredConfigPath: tsconfigPath,
        },
    );

    for (let i = 0; i < bundles.length; i++) {
        const bundle = bundles[i];
        const originalPath = relativeDeclarationPaths[i];

        const outputPath = resolve(postbundleOutDir, originalPath);

        writeFileSync(outputPath, bundle);
    }

    if (compilerOptions.declarationDir!.endsWith("dts-prebundle")) {
        rmSync(compilerOptions.declarationDir!, { recursive: true });
    }

    if (shouldDeleteTsConfig) {
        rmSync(tsconfigPath);
    }
}
