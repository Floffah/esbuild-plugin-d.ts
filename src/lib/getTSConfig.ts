import { readFileSync } from "fs";
import merge from "lodash.merge";
import { dirname, resolve } from "path";
import ts from "typescript";

export function getTSConfig(opts: {
    configPath?: string;
    configName?: string;
    searchPath?: string;
}) {
    let configPath =
        opts.configPath ??
        ts.findConfigFile(
            opts.searchPath ?? process.cwd(),
            ts.sys.fileExists,
            opts.configName,
        );
    if (!configPath) {
        throw new Error("No config file found");
    }

    if (configPath.startsWith(".")) {
        configPath = require.resolve(configPath);
    }

    const config = ts.readConfigFile(configPath, (path) =>
        readFileSync(path, "utf-8"),
    );

    if (config.config.extends) {
        const parentConfig = getTSConfig({
            ...opts,
            configPath: resolve(dirname(configPath), config.config.extends),
        });

        config.config = merge(config.config, parentConfig);
    }

    if (config.error) {
        throw config.error;
    } else {
        return config.config;
    }
}
