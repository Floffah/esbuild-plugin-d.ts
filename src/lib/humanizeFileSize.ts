export function humanizeFileSize(size: number): string {
    const i = Math.floor(Math.log(size) / Math.log(1024));
    return (
        Math.round((size / Math.pow(1024, i)) * 100) / 100 +
        ["b", "kb", "mb", "gb", "tb"][i]
    );
}
