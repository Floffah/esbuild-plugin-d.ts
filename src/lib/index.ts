export * from "./getCompilerOptions";
export * from "./resolveTSConfig";
export * from "./humanizeFileSize";
