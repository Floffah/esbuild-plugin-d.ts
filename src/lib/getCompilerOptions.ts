import { createHash } from "crypto";
import { BuildOptions } from "esbuild";
import { resolve } from "path";
import { tmpdir } from "tmp";
import ts from "typescript";

import { DTSPluginOpts } from "@/types";

export function getCompilerOptions(opts: {
    tsconfig: any;
    pluginOptions: DTSPluginOpts;
    esbuildOptions: BuildOptions;
}) {
    const compilerOptions = ts.convertCompilerOptionsFromJson(
        opts.tsconfig.compilerOptions,
        process.cwd(),
    ).options;

    compilerOptions.declaration = true;
    compilerOptions.emitDeclarationOnly = true;

    if (!compilerOptions.declarationDir) {
        compilerOptions.declarationDir =
            compilerOptions.declarationDir ??
            opts.esbuildOptions.outdir ??
            compilerOptions.outDir;
    }

    if (compilerOptions.incremental && !compilerOptions.tsBuildInfoFile) {
        const configHash = createHash("sha256")
            .update(
                JSON.stringify({
                    compilerOptions,
                    __buildContext: opts.pluginOptions?.__buildContext,
                }),
            )
            .digest("hex");

        compilerOptions.tsBuildInfoFile = resolve(
            opts.pluginOptions.buildInfoDir ?? tmpdir,
            `esbuild-plugin-dts-${configHash}.tsbuildinfo`,
        );
    }

    compilerOptions.listEmittedFiles = true;

    return compilerOptions;
}
