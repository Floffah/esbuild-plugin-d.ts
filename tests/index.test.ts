import { dtsPlugin, getTSConfig } from "../src";
import { getCompilerOptions } from "../src/lib/getCompilerOptions";
import { build } from "esbuild";
import { existsSync, readFileSync } from "fs";
import { rmSync } from "node:fs";
import { resolve } from "path";

const distDir = resolve(__dirname, "dist");

function readOutputFile() {
    return readFileSync(resolve(distDir, "index.d.ts"), "utf-8");
}

function clearDistDir() {
    if (existsSync(distDir)) {
        rmSync(distDir, { recursive: true });
    }
}

beforeEach(() => {
    clearDistDir();
});

test("Basic config", async () => {
    const tsconfig = resolve(__dirname, "./tsconfig.json");

    await build({
        plugins: [dtsPlugin({ tsconfig })],
        entryPoints: ["./tests/inputs/index.ts"],
        outdir: distDir,
        tsconfig,
    });

    expect(readOutputFile()).toMatchSnapshot();
});

test("Incremental mode", async () => {
    const tsconfig = resolve(__dirname, "./tsconfig.incremental.json");
    const __buildContext = Date.now();

    await build({
        plugins: [dtsPlugin({ tsconfig, __buildContext })],
        entryPoints: ["./tests/inputs/index.ts"],
        outdir: distDir,
        tsconfig,
    });

    expect(readOutputFile()).toMatchSnapshot();

    clearDistDir();

    await build({
        plugins: [dtsPlugin({ tsconfig, __buildContext })],
        entryPoints: ["./tests/inputs/index.ts"],
        outdir: distDir,
        tsconfig,
    });

    expect(existsSync(resolve(distDir, "index.d.ts"))).toBe(false);
});

test("TSConfig extends is resolved", async () => {
    const tsconfig = getTSConfig({
        configPath: resolve(__dirname, "./tsconfig.extends.json"),
    });

    const compilerOptions = getCompilerOptions({
        tsconfig,
        pluginOptions: {},
        esbuildOptions: {},
    });

    expect(compilerOptions).toMatchSnapshot();
    expect(compilerOptions.strict).toBe(true);
    expect(compilerOptions.emitDecoratorMetadata).toBe(true);
});
