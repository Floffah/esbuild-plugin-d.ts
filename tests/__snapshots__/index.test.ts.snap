// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic config 1`] = `
"export declare function someFunction(a: string, b: number): string;
export type SomeType = "someValue" | "anotherValue";
export declare const someVar: SomeType;
"
`;

exports[`Incremental mode 1`] = `
"export declare function someFunction(a: string, b: number): string;
export type SomeType = "someValue" | "anotherValue";
export declare const someVar: SomeType;
"
`;

exports[`TSConfig extends is resolved 1`] = `
{
  "allowSyntheticDefaultImports": true,
  "baseUrl": "/Users/ramstarh/Documents/Projects/plugin/esbuild-plugin-d.ts",
  "declaration": true,
  "declarationDir": "/Users/ramstarh/Documents/Projects/plugin/esbuild-plugin-d.ts/dist",
  "emitDeclarationOnly": true,
  "emitDecoratorMetadata": true,
  "esModuleInterop": true,
  "experimentalDecorators": true,
  "isolatedModules": true,
  "lib": [
    "lib.dom.d.ts",
    "lib.es2015.d.ts",
    "lib.es2017.d.ts",
    "lib.es2018.asynciterable.d.ts",
  ],
  "listEmittedFiles": true,
  "module": 1,
  "moduleResolution": 2,
  "noFallthroughCasesInSwitch": true,
  "noImplicitAny": true,
  "noImplicitReturns": true,
  "noImplicitThis": true,
  "noUnusedLocals": true,
  "noUnusedParameters": true,
  "outDir": "/Users/ramstarh/Documents/Projects/plugin/esbuild-plugin-d.ts/dist",
  "paths": {
    "@/*": [
      "./src/*",
    ],
  },
  "removeComments": true,
  "resolveJsonModule": true,
  "skipLibCheck": true,
  "sourceMap": true,
  "strict": true,
  "strictFunctionTypes": true,
  "strictNullChecks": true,
  "target": 2,
}
`;
